databricks_cli/oauth/oauth.py

```diff
@@ -0,0 +1,274 @@
+# Databricks CLI
+# Copyright 2021 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"), except
+# that the use of services to which certain application programming
+# interfaces (each, an "API") connect requires that the user first obtain
+# a license for the use of the APIs from Databricks, Inc. ("Databricks"),
+# by creating an account at www.databricks.com and agreeing to either (a)
+# the Community Edition Terms of Service, (b) the Databricks Terms of
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import base64
+import hashlib
+import json
+import os
+import webbrowser
+
+from datetime import datetime, timedelta, tzinfo
+
+import click
+
+import jwt
+from jwt import PyJWTError
+
+import oauthlib.oauth2
+from oauthlib.oauth2.rfc6749.errors import OAuth2Error
+
+import requests
+from requests.exceptions import RequestException
+
+from databricks_cli.utils import error_and_quit
+
+try:
+    from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+except ImportError:
+    from http.server import BaseHTTPRequestHandler, HTTPServer
+
+
+# This could use 'import secrets' in Python 3
+def token_urlsafe(nbytes=32):
+    tok = os.urandom(nbytes)
+    return base64.urlsafe_b64encode(tok).rstrip(b'=').decode('ascii')
+
+
+# This could be datetime.timezone.utc in Python 3
+class UTCTimeZone(tzinfo):
+    """UTC"""
+    def utcoffset(self, dt):
+        #pylint: disable=unused-argument
+        return timedelta(0)
+
+    def tzname(self, dt):
+        #pylint: disable=unused-argument
+        return "UTC"
+
+    def dst(self, dt):
+        #pylint: disable=unused-argument
+        return timedelta(0)
+
+
+# Some contant values
+OIDC_REDIRECTOR_PATH = "oidc"
+CLIENT_ID = "databricks-cli"
```

Comment:
For the UI side, each account corresponds to one app id. For cli part, seems we are using one app id to cover all the workspaces. Is my understanding correct?

Reply:
This is correct.